File: generate_navigation_diagrams.py
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches
from matplotlib.patches import FancyBboxPatch, Rectangle, Circle, Arrow
import numpy as np

# Configuration générale
plt.style.use('default')
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(20, 12))
fig.suptitle('Diagrammes de Navigation - FarmShop', fontsize=20, fontweight='bold', y=0.95)

# Couleurs personnalisées
colors = {
    'start': '#4CAF50',      # Vert
    'process': '#2196F3',    # Bleu
    'decision': '#FF9800',   # Orange
    'endpoint': '#9C27B0',   # Violet
    'payment': '#F44336',    # Rouge
    'success': '#8BC34A',    # Vert clair
    'arrow': '#666666'       # Gris
}

def add_rounded_box(ax, x, y, width, height, text, color, text_color='white'):
    """Ajouter une boîte arrondie avec du texte"""
    box = FancyBboxPatch((x-width/2, y-height/2), width, height,
                        boxstyle="round,pad=0.1", 
                        facecolor=color, 
                        edgecolor='black',
                        linewidth=1.5)
    ax.add_patch(box)
    ax.text(x, y, text, ha='center', va='center', 
           fontsize=9, fontweight='bold', color=text_color,
           wrap=True, bbox=dict(boxstyle="round,pad=0.3", alpha=0))

def add_decision_diamond(ax, x, y, text, color, text_color='white'):
    """Ajouter un losange de décision"""
    diamond_points = np.array([[x, y+0.8], [x+1.2, y], [x, y-0.8], [x-1.2, y]])
    diamond = plt.Polygon(diamond_points, facecolor=color, edgecolor='black', linewidth=1.5)
    ax.add_patch(diamond)
    ax.text(x, y, text, ha='center', va='center', 
           fontsize=8, fontweight='bold', color=text_color)

def add_arrow(ax, start_x, start_y, end_x, end_y, text='', offset=0.3):
    """Ajouter une flèche avec du texte optionnel"""
    ax.annotate('', xy=(end_x, end_y), xytext=(start_x, start_y),
                arrowprops=dict(arrowstyle='->', lw=2, color=colors['arrow']))
    if text:
        mid_x, mid_y = (start_x + end_x) / 2, (start_y + end_y) / 2 + offset
        ax.text(mid_x, mid_y, text, ha='center', va='center', 
               fontsize=7, style='italic', 
               bbox=dict(boxstyle="round,pad=0.2", facecolor='white', alpha=0.8))

# ============= DIAGRAMME DE NAVIGATION - ACHAT =============
ax1.set_title('Navigation - Processus d\'Achat', fontsize=16, fontweight='bold', pad=20)

# Points de navigation pour l'achat
# Page d'accueil / Catalogue
add_rounded_box(ax1, 2, 18, 2.5, 1, 'Page d\'Accueil\n/products', colors['start'])

# Recherche / Filtrage
add_rounded_box(ax1, 2, 16, 2.5, 1, 'Recherche/Filtrage\nProduits', colors['process'])

# Page produit
add_rounded_box(ax1, 2, 14, 2.5, 1, 'Détail Produit\n/products/{slug}', colors['process'])

# Décision: Ajouter au panier
add_decision_diamond(ax1, 2, 12, 'Ajouter au\npanier?', colors['decision'])

# Panier
add_rounded_box(ax1, 6, 14, 2.5, 1, 'Panier\n/cart', colors['process'])

# Connexion/Inscription
add_decision_diamond(ax1, 6, 12, 'Utilisateur\nconnecté?', colors['decision'])

add_rounded_box(ax1, 10, 12, 2.5, 1, 'Connexion/Inscription\n/login /register', colors['process'])

# Checkout
add_rounded_box(ax1, 6, 10, 2.5, 1, 'Checkout\n/checkout', colors['process'])

# Validation commande
add_rounded_box(ax1, 6, 8, 2.5, 1, 'Validation\nCommande', colors['process'])

# Paiement
add_rounded_box(ax1, 6, 6, 2.5, 1, 'Paiement Stripe\n/payment/{order}', colors['payment'])

# Décision paiement
add_decision_diamond(ax1, 6, 4, 'Paiement\nréussi?', colors['decision'])

# Succès
add_rounded_box(ax1, 10, 4, 2.5, 1, 'Confirmation\n/orders/{order}', colors['success'])

# Échec
add_rounded_box(ax1, 2, 4, 2.5, 1, 'Échec Paiement\n/payment-failed', colors['payment'])

# Mes commandes
add_rounded_box(ax1, 10, 2, 2.5, 1, 'Mes Commandes\n/orders', colors['endpoint'])

# Flèches de navigation - Achat
add_arrow(ax1, 2, 17.5, 2, 16.5)  # Accueil -> Recherche
add_arrow(ax1, 2, 15.5, 2, 14.5)  # Recherche -> Produit
add_arrow(ax1, 2, 13.2, 2, 12.8)  # Produit -> Décision panier
add_arrow(ax1, 3.2, 12, 4.8, 13.5, 'Oui')  # Décision -> Panier
add_arrow(ax1, 6, 13.5, 6, 12.8)  # Panier -> Décision connexion
add_arrow(ax1, 7.2, 12, 8.8, 12, 'Non')  # Décision -> Connexion
add_arrow(ax1, 10, 11.5, 6.5, 10.8, 'Après connexion')  # Connexion -> Checkout
add_arrow(ax1, 6, 11.5, 6, 10.5, 'Oui')  # Décision connexion -> Checkout
add_arrow(ax1, 6, 9.5, 6, 8.5)  # Checkout -> Validation
add_arrow(ax1, 6, 7.5, 6, 6.5)  # Validation -> Paiement
add_arrow(ax1, 6, 5.5, 6, 4.8)  # Paiement -> Décision
add_arrow(ax1, 7.2, 4, 8.8, 4, 'Oui')  # Décision -> Succès
add_arrow(ax1, 4.8, 4, 3.2, 4, 'Non')  # Décision -> Échec
add_arrow(ax1, 10, 3.5, 10, 2.5)  # Succès -> Mes commandes
add_arrow(ax1, 0.8, 12, 0.8, 4.5, '')  # Retour depuis décision panier
add_arrow(ax1, 1, 4.5, 2, 4.5, 'Retry')  # Échec -> Retour

# Configuration de l'axe pour l'achat
ax1.set_xlim(-1, 12)
ax1.set_ylim(0, 20)
ax1.set_aspect('equal')
ax1.axis('off')

# ============= DIAGRAMME DE NAVIGATION - LOCATION =============
ax2.set_title('Navigation - Processus de Location', fontsize=16, fontweight='bold', pad=20)

# Points de navigation pour la location
# Page d'accueil / Catalogue location
add_rounded_box(ax2, 2, 18, 2.5, 1, 'Catalogue Location\n/rentals', colors['start'])

# Filtrage par catégorie
add_rounded_box(ax2, 2, 16, 2.5, 1, 'Filtrage Équipements\nDisponibles', colors['process'])

# Page produit location
add_rounded_box(ax2, 2, 14, 2.5, 1, 'Détail Équipement\n/rentals/{slug}', colors['process'])

# Vérification disponibilité
add_decision_diamond(ax2, 2, 12, 'Équipement\ndisponible?', colors['decision'])

# Panier location
add_rounded_box(ax2, 6, 14, 2.5, 1, 'Panier Location\n/cart-location', colors['process'])

# Sélection dates
add_rounded_box(ax2, 6, 12, 2.5, 1, 'Sélection Dates\nLocation', colors['process'])

# Connexion pour location
add_decision_diamond(ax2, 6, 10, 'Utilisateur\nconnecté?', colors['decision'])

add_rounded_box(ax2, 10, 10, 2.5, 1, 'Connexion Requise\n/login', colors['process'])

# Checkout location
add_rounded_box(ax2, 6, 8, 2.5, 1, 'Checkout Location\n/checkout-rental', colors['process'])

# Validation location
add_rounded_box(ax2, 6, 6.5, 2.5, 1, 'Validation\n+ Caution', colors['process'])

# Paiement location
add_rounded_box(ax2, 6, 5, 2.5, 1, 'Paiement + Caution\n/payment-rental/{order}', colors['payment'])

# Décision paiement location
add_decision_diamond(ax2, 6, 3.5, 'Paiement\nréussi?', colors['decision'])

# Succès location
add_rounded_box(ax2, 10, 3.5, 2.5, 1, 'Location Confirmée\n/rental-orders/{order}', colors['success'])

# Suivi location
add_rounded_box(ax2, 10, 2, 2.5, 1, 'Mes Locations\n/rental-orders', colors['endpoint'])

# Gestion retour
add_rounded_box(ax2, 10, 0.5, 2.5, 1, 'Retour Équipement\n+ Inspection', colors['endpoint'])

# Échec location
add_rounded_box(ax2, 2, 3.5, 2.5, 1, 'Échec Paiement\n/payment-failed', colors['payment'])

# Flèches de navigation - Location
add_arrow(ax2, 2, 17.5, 2, 16.5)  # Accueil -> Filtrage
add_arrow(ax2, 2, 15.5, 2, 14.5)  # Filtrage -> Produit
add_arrow(ax2, 2, 13.2, 2, 12.8)  # Produit -> Décision dispo
add_arrow(ax2, 3.2, 12, 4.8, 13.5, 'Oui')  # Décision -> Panier location
add_arrow(ax2, 6, 13.5, 6, 12.5)  # Panier -> Dates
add_arrow(ax2, 6, 11.5, 6, 10.8)  # Dates -> Décision connexion
add_arrow(ax2, 7.2, 10, 8.8, 10, 'Non')  # Décision -> Connexion
add_arrow(ax2, 10, 9.5, 6.5, 8.8, 'Après connexion')  # Connexion -> Checkout
add_arrow(ax2, 6, 9.2, 6, 8.5, 'Oui')  # Décision connexion -> Checkout
add_arrow(ax2, 6, 7.5, 6, 7)  # Checkout -> Validation
add_arrow(ax2, 6, 6, 6, 5.5)  # Validation -> Paiement
add_arrow(ax2, 6, 4.5, 6, 4.3)  # Paiement -> Décision
add_arrow(ax2, 7.2, 3.5, 8.8, 3.5, 'Oui')  # Décision -> Succès
add_arrow(ax2, 4.8, 3.5, 3.2, 3.5, 'Non')  # Décision -> Échec
add_arrow(ax2, 10, 3, 10, 2.5)  # Succès -> Mes locations
add_arrow(ax2, 10, 1.5, 10, 1)  # Mes locations -> Retour
add_arrow(ax2, 0.8, 12, 0.8, 4, '')  # Retour depuis décision dispo
add_arrow(ax2, 1, 4, 2, 4, 'Non disponible')  # Indisponible -> Échec

# Configuration de l'axe pour la location
ax2.set_xlim(-1, 12)
ax2.set_ylim(-1, 20)
ax2.set_aspect('equal')
ax2.axis('off')

# Légende commune
legend_elements = [
    mpatches.Rectangle((0, 0), 1, 1, facecolor=colors['start'], label='Point de départ'),
    mpatches.Rectangle((0, 0), 1, 1, facecolor=colors['process'], label='Processus'),
    mpatches.Rectangle((0, 0), 1, 1, facecolor=colors['decision'], label='Décision'),
    mpatches.Rectangle((0, 0), 1, 1, facecolor=colors['payment'], label='Paiement'),
    mpatches.Rectangle((0, 0), 1, 1, facecolor=colors['success'], label='Succès'),
    mpatches.Rectangle((0, 0), 1, 1, facecolor=colors['endpoint'], label='Point final')
]

fig.legend(handles=legend_elements, loc='lower center', ncol=6, 
          bbox_to_anchor=(0.5, 0.02), fontsize=12)

plt.tight_layout()
plt.subplots_adjust(bottom=0.1)
plt.savefig('diagrammes_navigation_farmshop.png', dpi=300, bbox_inches='tight', 
           facecolor='white', edgecolor='none')
plt.show()

print("Diagrammes de navigation générés avec succès : diagrammes_navigation_farmshop.png")

File: app/Console/Commands/GenerateNavigationDiagrams.php
<?php

namespace App\Console\Commands;

use Illuminate\Console\Command;

class GenerateNavigationDiagrams extends Command
{
    /**
     * The name and signature of the console command.
     *
     * @var string
     */
    protected $signature = 'farmshop:generate-navigation-diagrams';

    /**
     * The console command description.
     *
     * @var string
     */
    protected $description = 'Génère les diagrammes de navigation pour les processus d\'achat et de location';

    /**
     * Execute the console command.
     */
    public function handle()
    {
        $this->info('=== Génération des Diagrammes de Navigation FarmShop ===');
        $this->newLine();

        try {
            // Configuration du gestionnaire d'images
            $manager = new ImageManager(new Driver());

            // Créer l'image de base
            $width = 1600;
            $height = 1200;
            $image = $manager->create($width, $height);

            // Couleurs
            $colors = [
                'background' => '#ffffff',
                'start' => '#4CAF50',      // Vert
                'process' => '#2196F3',    // Bleu
                'decision' => '#FF9800',   // Orange
                'endpoint' => '#9C27B0',   // Violet
                'payment' => '#F44336',    // Rouge
                'success' => '#8BC34A',    // Vert clair
                'text' => '#000000',       // Noir
                'arrow' => '#666666'       // Gris
            ];

            // Remplir le fond
            $image->fill($colors['background']);

            $this->info('📊 Création du canvas de base...');

            // Titre principal
            $image->text('DIAGRAMMES DE NAVIGATION FARMSHOP', $width/2, 40, function ($font) {
                $font->size(24);
                $font->color('#000000');
                $font->align('center');
                $font->weight('bold');
            });

            // ========== DIAGRAMME ACHAT ==========
            $this->info('🛒 Génération du diagramme d\'achat...');

            // Titre achat
            $image->text('PROCESSUS D\'ACHAT', $width/4, 80, function ($font) {
                $font->size(18);
                $font->color('#000000');
                $font->align('center');
                $font->weight('bold');
            });

            // Éléments du processus d'achat (côté gauche)
            $leftX = $width/4;
            $boxWidth = 160;
            $boxHeight = 50;

            $achatSteps = [
                ['y' => 140, 'text' => "Page d'Accueil\n/products", 'color' => $colors['start']],
                ['y' => 210, 'text' => "Catalogue\nProduits", 'color' => $colors['process']],
                ['y' => 280, 'text' => "Détail Produit\n/products/{slug}", 'color' => $colors['process']],
                ['y' => 350, 'text' => "Panier\n/cart", 'color' => $colors['process']],
                ['y' => 420, 'text' => "Connexion\n/login", 'color' => $colors['decision']],
                ['y' => 490, 'text' => "Checkout\n/checkout", 'color' => $colors['process']],
                ['y' => 560, 'text' => "Paiement Stripe\n/payment/{order}", 'color' => $colors['payment']],
                ['y' => 630, 'text' => "Confirmation\n/orders/{order}", 'color' => $colors['success']],
                ['y' => 700, 'text' => "Mes Commandes\n/orders", 'color' => $colors['endpoint']]
            ];

            foreach ($achatSteps as $index => $step) {
                $this->drawRoundedBox($image, $leftX, $step['y'], $boxWidth, $boxHeight, $step['text'], $step['color']);
                
                // Dessiner la flèche vers l'étape suivante
                if ($index < count($achatSteps) - 1) {
                    $this->drawArrow($image, $leftX, $step['y'] + $boxHeight/2, $leftX, $achatSteps[$index + 1]['y'] - $boxHeight/2);
                }
            }

            // ========== DIAGRAMME LOCATION ==========
            $this->info('🚜 Génération du diagramme de location...');

            // Titre location
            $image->text('PROCESSUS DE LOCATION', 3*$width/4, 80, function ($font) {
                $font->size(18);
                $font->color('#000000');
                $font->align('center');
                $font->weight('bold');
            });

            // Éléments du processus de location (côté droit)
            $rightX = 3*$width/4;

            $locationSteps = [
                ['y' => 140, 'text' => "Catalogue Location\n/rentals", 'color' => $colors['start']],
                ['y' => 210, 'text' => "Filtrage\nÉquipements", 'color' => $colors['process']],
                ['y' => 280, 'text' => "Détail Équipement\n/rentals/{slug}", 'color' => $colors['process']],
                ['y' => 350, 'text' => "Panier Location\n/cart-location", 'color' => $colors['process']],
                ['y' => 420, 'text' => "Sélection Dates\nLocation", 'color' => $colors['decision']],
                ['y' => 490, 'text' => "Checkout Location\n/checkout-rental", 'color' => $colors['process']],
                ['y' => 560, 'text' => "Paiement + Caution\n/payment-rental", 'color' => $colors['payment']],
                ['y' => 630, 'text' => "Location Confirmée\n/rental-orders/{order}", 'color' => $colors['success']],
                ['y' => 700, 'text' => "Mes Locations\n/rental-orders", 'color' => $colors['endpoint']],
                ['y' => 770, 'text' => "Retour Équipement\n+ Inspection", 'color' => $colors['endpoint']]
            ];

            foreach ($locationSteps as $index => $step) {
                $this->drawRoundedBox($image, $rightX, $step['y'], $boxWidth, $boxHeight, $step['text'], $step['color']);
                
                // Dessiner la flèche vers l'étape suivante
                if ($index < count($locationSteps) - 1) {
                    $this->drawArrow($image, $rightX, $step['y'] + $boxHeight/2, $rightX, $locationSteps[$index + 1]['y'] - $boxHeight/2);
                }
            }

            // Ligne de séparation
            $image->drawLine(function ($draw) use ($width, $height) {
                $draw->from($width/2, 100);
                $draw->to($width/2, $height - 100);
                $draw->color('#cccccc');
                $draw->width(3);
            });

            // ========== LÉGENDE ==========
            $this->info('🎨 Ajout de la légende...');

            $legendY = $height - 100;
            $image->text('LÉGENDE', $width/2, $legendY - 50, function ($font) {
                $font->size(16);
                $font->color('#000000');
                $font->align('center');
                $font->weight('bold');
            });

            $legendItems = [
                ['color' => $colors['start'], 'text' => 'Départ'],
                ['color' => $colors['process'], 'text' => 'Processus'],
                ['color' => $colors['decision'], 'text' => 'Décision'],
                ['color' => $colors['payment'], 'text' => 'Paiement'],
                ['color' => $colors['success'], 'text' => 'Succès'],
                ['color' => $colors['endpoint'], 'text' => 'Fin']
            ];

            $legendStartX = $width/2 - 300;
            foreach ($legendItems as $index => $item) {
                $x = $legendStartX + ($index * 100);
                
                // Carré de couleur
                $image->drawRectangle($x - 15, $legendY - 10, function ($draw) use ($item) {
                    $draw->background($item['color']);
                    $draw->border(2, '#000000');
                });
                
                // Texte
                $image->text($item['text'], $x + 25, $legendY, function ($font) {
                    $font->size(12);
                    $font->color('#000000');
                    $font->align('left');
                    $font->valign('middle');
                });
            }

            // Sauvegarder l'image
            $filename = public_path('storage/diagrammes_navigation_farmshop.png');
            $image->save($filename);

            $this->info("✅ Diagramme généré avec succès !");
            $this->info("📁 Fichier : {$filename}");
            $this->info("📊 Résolution : {$width}x{$height} pixels");
            $this->newLine();

            $this->comment("📦 ACHAT : Accueil → Catalogue → Produit → Panier → Connexion → Checkout → Paiement → Confirmation → Commandes");
            $this->comment("🚜 LOCATION : Catalogue → Filtrage → Équipement → Panier → Dates → Checkout → Paiement+Caution → Confirmation → Suivi → Retour");

        } catch (\Exception $e) {
            $this->error("❌ Erreur lors de la génération : " . $e->getMessage());
            return 1;
        }

        return 0;
    }

    /**
     * Dessine une boîte arrondie avec texte
     */
    private function drawRoundedBox($image, $x, $y, $width, $height, $text, $bgColor, $textColor = '#ffffff')
    {
        // Dessiner le rectangle
        $image->drawRectangle($x - $width/2, $y - $height/2, function ($draw) use ($bgColor, $width, $height) {
            $draw->background($bgColor);
            $draw->border(2, '#000000');
        });
        
        // Ajouter le texte
        $image->text($text, $x, $y, function ($font) use ($textColor) {
            $font->size(11);
            $font->color($textColor);
            $font->align('center');
            $font->valign('middle');
        });
    }

    /**
     * Dessine une flèche entre deux points
     */
    private function drawArrow($image, $startX, $startY, $endX, $endY)
    {
        // Dessiner la ligne
        $image->drawLine(function ($draw) use ($startX, $startY, $endX, $endY) {
            $draw->from($startX, $startY);
            $draw->to($endX, $endY);
            $draw->color('#666666');
            $draw->width(2);
        });
        
        // Dessiner la pointe de la flèche
        $angle = atan2($endY - $startY, $endX - $startX);
        $arrowLength = 8;
        $arrowAngle = pi() / 6;
        
        $x1 = $endX - $arrowLength * cos($angle - $arrowAngle);
        $y1 = $endY - $arrowLength * sin($angle - $arrowAngle);
        $x2 = $endX - $arrowLength * cos($angle + $arrowAngle);
        $y2 = $endY - $arrowLength * sin($angle + $arrowAngle);
        
        $image->drawLine(function ($draw) use ($endX, $endY, $x1, $y1) {
            $draw->from($endX, $endY);
            $draw->to($x1, $y1);
            $draw->color('#666666');
            $draw->width(2);
        });
        
        $image->drawLine(function ($draw) use ($endX, $endY, $x2, $y2) {
            $draw->from($endX, $endY);
            $draw->to($x2, $y2);
            $draw->color('#666666');
            $draw->width(2);
        });
    }
}
